### PR TITLE
Fix an error when an exercise has no due date defined

### PR DESCRIPTION
**Pull request by FrankeLukas**
### Motivation and Context

There are some TypeErrors like `o.isBefore is not a function. (In 'o.isBefore(z.now())', 'o.isBefore' is undefined)` on Sentry.

### Description

We now check if the due date is null/undefined before calling `isBefore` on it. From https://github.com/ls1intum/Artemis/pull/3394#issuecomment-837368116:
> I think the problem is that `date?.isBefore(xyz)` with `date == undefined` evaluates to `undefined?.isBefore(xyz)` and then `undefined(xyz)` (which obviously doesn't work). If I understand https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining correctly we would have to use `date?.isBefore?.(xyz)` instead (which I didn't know before and) which we don't use in the code base afaik?

### Steps for Testing

I actually don't know how to trigger this error, but the code changes alone should be indication enough that it is fixed. (Only other option I see is that we're trying to call `isBefore` on an object that isn't actually a date, but I don't know how that would happen.)

### CODE REVIEW

**Review by rs-maker — DISMISSED** [review]
Optional chaining with 
'foo?.bar'
should only call the .bar function if foo is not null anyways
<a href="https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html"> optional chaining </a>
maybe 
"changedParticipation.exercise?.dueDate?.isBefore(moment.now()) ?? false ... "
is a bit shorter and gets the same job done but your version works too so i approve anyways.

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts [hunk 1]
```diff
@@ -263,7 +263,7 @@ export class CourseExerciseDetailsComponent implements OnInit, OnDestroy {
         this.participationUpdateListener = this.participationWebsocketService.subscribeForParticipationChanges().subscribe((changedParticipation: StudentParticipation) => {
             if (changedParticipation && this.exercise && changedParticipation.exercise?.id === this.exercise.id) {
                 // Notify student about late submission result
-                if (changedParticipation.exercise?.dueDate?.isBefore(moment.now()) && changedParticipation.results?.length! > this.studentParticipation?.results?.length!) {
+                if (changedParticipation.exercise?.dueDate && changedParticipation.exercise!.dueDate!.isBefore(moment.now()) && changedParticipation.results?.length! > this.studentParticipation?.results?.length!) {
```
Comment (FuchsDominik) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:266]:
```suggestion
                if (changedParticipation.exercise?.dueDate &&
                changedParticipation.exercise!.dueDate!.isBefore(moment.now()) &&
                changedParticipation.results?.length! > this.studentParticipation?.results?.length!) {
```
code-style issues

**Review by FuchsDominik — DISMISSED** [review]
Apart from the code style looks good, I don't quite understand why isBefore() is called despite the optional chaning like Robert already mentioned, but your fix should make it work

**Review by FuchsDominik — APPROVED** [review]

**Review by ivohashamov — APPROVED** [review]
Looks good

**Comment by FrankeLukas** [comment]
@rs-maker @FuchsDominik I think the problem is that `date?.isBefore(xyz)` with `date == undefined` evaluates to `undefined?.isBefore(xyz)` and then `undefined(xyz)` (which obviously doesn't work). If I understand https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining correctly we would have to use `date?.isBefore?.(xyz)` instead (which I didn't know before and) which we don't use in the code base afaik?

**Comment by iozbeyli** [comment]
> @rs-maker @FuchsDominik I think the problem is that `date?.isBefore(xyz)` with `date == undefined` evaluates to `undefined?.isBefore(xyz)` and then `undefined(xyz)` (which obviously doesn't work). If I understand https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining correctly we would have to use `date?.isBefore?.(xyz)` instead (which I didn't know before and) which we don't use in the code base afaik?

That is true if date is evaluating to something else that does not have an isBefore method. I don't think it evaluates undefined(xyz) when date is undefined it just returns undefined.

**Pull request merged**